RFC/0001-graph-display.md

```diff
@@ -0,0 +1,241 @@
+(e.g. allowing different colours to be chosen in a dark mode).
+
+### 3.4 Implementation
+
+In the actual QML, we will create a `GraphNodeView` that takes two
+parameters.  First first parameter is the `AbstractGraphModel` that
+contains the graph information.  The second is a delegate that accepts
+the `NodeInterface` and returns a group of widgets.  This delegate
+will *not* be responsible for adding the title bar to the node or
+drawing the node border.
+
+The implementation of this QML will be two `Repeaters`.  The first
+will iterate over the `nodes`, create the header and border, and use
+the provided delegate to populate the nodes.  The second will iterate
+over the `edges` and draw the splines.
+
+
+## 4 Metrics & Dashboards
+
+The final implementation will need to provide the following abilities:
+
+- Add new nodes of any type
+- Drag nodes to a new position
+- Draw connections between nodes with the mouse
+- Delete connections
+- Auto-arrange nodes
+- Display the nodes at multiple zoom levels
+- Scroll around the node environment
+
+None of these actions should leave the graph in an invalid state.
+
+## 5 Drawbacks
+
+- This project provides a significant time sink for at least one developer.
+
+- Even once the general library is written, more developer time will
+  be needed to write the Dissolve specific uses of the library
+
+
+## 6 Alternatives
+
+[Qt Node Editor](https://github.com/paceholder/nodeeditor) does not
+natively support QML and thus will require significant work to provide
+a consistent experience with the rest of the application.
+
+[QuickQanava](https://github.com/cneben/QuickQanava?tab=readme-ov-file)
+requires the construction and maintenance of an explicit graph object,
+as opposed to providing an abstract interface that we can build our
+own model again.  Additionally, at the time of this writing,
+QuickQanava does not build on Mac or Linux
+
+## 7 Potential Impact and Dependencies
+
+This could have a significant impact on the structure of the input
+file, since it represents a major change in the state of the data.
+Furthermore, this may be a significant break change, as it would not
+be a trivial matter to convert existing, linear code into the new
+graph based setup.
+
+## 8 Unresolved questions
+  
+### 8.1 Invalid Graphs
+
+Users might attempt to connect nodes in incompatible way
+(e.g. connecting a string to a value expecting an int) or fail to
+provide a node with all of the information that it requires.  More
+generally, the system needs to deal with the possibility of invalid
+graphs.
+  
+This first method is to simply forbid invalid graphs.  The model
+should reject any attempt to create invalid edges or delete nodes that
+are required.  This had advantages in design and implementation, but
+places limits on what can provide a good user experience.  For
+example, every possible node input must have a **valid** default
+value, so that the connection to that input can be deleted.
```

Comment:
I think that strictly disallowing connections between dissimilar input / output types is the best route to go - something like Blender will allow you to connect related type inputs/outputs and still yield sensible results, but I don't think we need that softness here. Regarding the need for default values, this is what we currently have with keywords in modules and nodes, so not a problem moving forward.